pkg/identityserver/gormstore/end_device_store.go

```diff
@@ -178,6 +178,12 @@ func (s *deviceStore) UpdateEndDevice(ctx context.Context, dev *ttnpb.EndDevice,
 		return nil, err
 	}
 	oldAttributes, oldLocations, oldPicture := devModel.Attributes, devModel.Locations, devModel.Picture
+	if ttnpb.HasAnyField(fieldMask, "last_seen_at") && dev.GetLastSeenAt() != nil {
+		timestamp := ttnpb.StdTime(dev.GetLastSeenAt())
+		if devModel.LastSeenAt != nil && timestamp.Before(*devModel.LastSeenAt) {
+			dev.LastSeenAt = ttnpb.ProtoTime(devModel.LastSeenAt)
+		}
+	}
```

Comment:
The problem (with this implementation as well) is that it could race (multiple AS instances updating at the same time). 
Now that I am thinking I guess we don't really care about these timestamps being the most accurate (in case of race it will be a ms difference, I suppose). So I think it would be acceptable to have some data races here?

Reply:
Even the same Application Server will race with itself in the model that we have. It is fine to have eventual consistency.